### ISSUE TITLE
Logging comparison

### ISSUE THREAD

**Issue by Danp2** [Danp2/au3WebDriver#316]
PR https://github.com/Danp2/au3WebDriver/pull/290/ has proposed changes to the UDF logging. This issue is a place to discuss the changes and determine any needed changes to the PR.

**Comment by Danp2** [comment]
The following represent the output for a call to _WD_Navigate --

### Current Info
```
__WD_Post: URL=HTTP://127.0.0.1:4444/session/c9b422d8-f1d5-49a0-9253-617487798318/url; $sData={"url":"https://google.com"}
__WD_Post ==> Success (0) HTTP status = 200 : ResponseText={"value":null}...
_WD_Navigate ==> Success (0) HTTP status = 200 : {"value":null}
```

### Current Full

```
__WD_Post: URL=HTTP://127.0.0.1:4444/session/0d1040d0-4e85-47ae-9d4d-77071450c0cc/url; $sData={"url":"https://google.com"}
__WD_Post ==> Success (0) HTTP status = 200 : ResponseText={"value":null}...
_WD_Navigate ==> Success (0) HTTP status = 200 : {"value":null}
```

### New Info

```
__WD_Post ==> Success [0] : HTTP status = 200
_WD_Navigate ==> Success [0]
```

### New Full

```
__WD_Post: URL=HTTP://127.0.0.1:4444/session/ab17d6e3-fb50-42da-a1ca-9867feb7d931/url; $sData={"url":"https://google.com"}
__WD_Post ==> Success [0] : HTTP status = 200 ResponseText={"value":null}
_WD_Navigate ==> Success [0]
```

**Comment by Danp2** [comment]
Looking at the above, IMO the `New Info` section is lacking in details.

**Comment by Danp2** [comment]
Here's another example, but with a failed navigation attempt --

### Current Info

```
__WD_Post: URL=HTTP://127.0.0.1:4444/session/78ea70b2-cb4b-49b1-b772-e85c83bc25b2/url; $sData={"url":"www.htmlquick.com/reference/tags/input-file.html"}
__WD_Post ==> Invalid argument (5) HTTP status = 400 : ResponseText={"value":{"error":"invalid argument","message":"Malformed URL: URL constructor: www.htmlquick.com/re...
_WD_Navigate ==> Invalid argument (5) HTTP status = 400 : {"value":{"error":"invalid argument","message":"Malformed URL: URL constructor: www.htmlquick.com/reference/tags/input-file.html is not a valid URL.","stacktrace":"WebDriverError@chrome://remote/content/shared/webdriver/Errors.jsm:183:5\nInvalidArgumentError@chrome://remote/content/shared/webdriver/Errors.jsm:312:5\nGeckoDriver.prototype.navigateTo@chrome://remote/content/marionette/driver.js:791:11\n"}}
```

### Current Full

```
__WD_Post: URL=HTTP://127.0.0.1:4444/session/fcdfb2a6-04b0-44fe-b695-863035482a23/url; $sData={"url":"www.htmlquick.com/reference/tags/input-file.html"}
__WD_Post ==> Invalid argument (5) HTTP status = 400 : ResponseText={"value":{"error":"invalid argument","message":"Malformed URL: URL constructor: www.htmlquick.com/re...
_WD_Navigate ==> Invalid argument (5) HTTP status = 400 : {"value":{"error":"invalid argument","message":"Malformed URL: URL constructor: www.htmlquick.com/reference/tags/input-file.html is not a valid URL.","stacktrace":"WebDriverError@chrome://remote/content/shared/webdriver/Errors.jsm:183:5\nInvalidArgumentError@chrome://remote/content/shared/webdriver/Errors.jsm:312:5\nGeckoDriver.prototype.navigateTo@chrome://remote/content/marionette/driver.js:791:11\n"}}
```

### New Info

```
__WD_Post ==> Invalid argument [5] : HTTP status = 400 ResponseText=...
_WD_Navigate ==> Invalid argument [5]
```

### New Full

```
__WD_Post: URL=HTTP://127.0.0.1:4444/session/5a170b13-6ccf-4b6d-80d7-66903c1f6976/url; $sData={"url":"www.htmlquick.com/reference/tags/input-file.html"}
__WD_Post ==> Invalid argument [5] : HTTP status = 400 ResponseText={"value":{"error":"invalid argument","message":"Malformed URL: URL constructor: www.htmlquick.com/reference/tags/input-file.html is not a valid URL.","stacktrace":"WebDriverError@chrome://remote/content/shared/webdriver/Errors.jsm:183:5\nInvalidArgumentError@chrome://remote/content/shared/webdriver/Errors.jsm:312:5\nGeckoDriver.prototype.navigateTo@chrome://remote/content/marionette/driver.js:791:11\n"}}
_WD_Navigate ==> Invalid argument [5]
```

**Comment by Danp2** [comment]
Need to figure out why the ResponseText is missing under `New Info` above.

**Comment by mlipok** [comment]
Please edit examples by just reordering them

Current info
New info

Current full
New full

**Comment by Danp2** [comment]
I'm evaluating Info vs Full in addition to Current vs New, so the current order works for me. You can simply copy/paste into the program of your choice if you want to reorder for comparison purposes.

**Comment by mlipok** [comment]
> Need to figure out why the ResponseText is missing under `New Info` above.

It is intentionall

**Comment by Danp2** [comment]
> It is intentionall

If that is true, then why show `ResponseText=...`? 😦

**Comment by mlipok** [comment]
I will investigate tomorrow.

**Comment by Danp2** [comment]
I think we also need to review this comment and the associated change -- https://github.com/Danp2/au3WebDriver/pull/290#discussion_r873213500

IMO, both `$_WD_DEBUG_Info` and `$_WD_DEBUG_Full` debugging levels should support trimming of the response if that is what the script author has configured with $_WD_RESPONSE_TRIM.

The above change completely eliminated the display of the response text, even if an error has occurred.

**Comment by mlipok** [comment]
> > It is intentionall
> 
> If that is true, then why show `ResponseText=...`? 😦

https://github.com/Danp2/au3WebDriver/pull/290#issuecomment-1098538452

**Comment by mlipok** [comment]
I have some rethinking, but I laying in the bed, writing on smarphone.
Will back tomorrow.

**Comment by Danp2** [comment]
> https://github.com/Danp2/au3WebDriver/pull/290#issuecomment-1098538452

That discussion was continued in https://github.com/Danp2/au3WebDriver/issues/280#issuecomment-1099279080

> Please remember that: $_WD_RESPONSE_TRIM = -1

The goal was that -1 would result in untrimmed text being returned. 

> Need to figure out why the ResponseText is missing under New Info above.

This is coming from __WD_Post, where the updates haven't been applied. Thus, no real issue at the moment. 🙂

**Comment by mlipok** [comment]
https://github.com/Danp2/au3WebDriver/issues/280#issuecomment-1099993643
> My idea was to completely remove these lines unless $_WD_DEBUG_Full was in use.

**Comment by mlipok** [comment]
Working on a fix

**Comment by mlipok** [comment]
Take a look on:
[__WD_MessageCreator](https://github.com/Danp2/au3WebDriver/pull/290/commits/99406453dd1c8852b82d61fdd9287cc78c4b482c)

**Comment by Danp2** [comment]
This is an updated example log --

### New Info

```
+ wd_demo.au3: Running: UserTesting
__WD_Post: URL=HTTP://127.0.0.1:4444/session/15607a78-2c2b-4a2d-8701-89eee936182c/url
__WD_Post ==> Success [0] : HTTP status = 200
_WD_Navigate ==> Success [0]
__WD_Post: URL=HTTP://127.0.0.1:4444/session/15607a78-2c2b-4a2d-8701-89eee936182c/execute/sync
__WD_Post ==> Success [0] : HTTP status = 200
_WD_ExecuteScript ==> Success [0]
_WD_LoadWait ==> Success [0]
__WD_Post: URL=HTTP://127.0.0.1:4444/session/15607a78-2c2b-4a2d-8701-89eee936182c/element
__WD_Post ==> Invalid Expression [11] : HTTP status = 400
_WD_FindElement ==> Invalid Expression [11] : Parameters:   Strategy=xpath   Selector=
__WD_Post: URL=HTTP://127.0.0.1:4444/session/15607a78-2c2b-4a2d-8701-89eee936182c/element
__WD_Post ==> Invalid Expression [11] : HTTP status = 400
_WD_FindElement ==> Invalid Expression [11] : Parameters:   Strategy=css selector   Selector=
+ wd_demo.au3: Finished: UserTesting
```

### New Full

```
+ wd_demo.au3: Running: UserTesting
__WD_Post: URL=HTTP://127.0.0.1:4444/session/f8e62cdd-0833-483c-aec5-ebac4e0bd7a4/url; $sData={"url":"https://www.google.com"}
__WD_Post: URL=HTTP://127.0.0.1:4444/session/f8e62cdd-0833-483c-aec5-ebac4e0bd7a4/url
__WD_Post ==> Success [0] : HTTP status = 200 ResponseText={"value":null}
_WD_Navigate ==> Success [0]
__WD_Post: URL=HTTP://127.0.0.1:4444/session/f8e62cdd-0833-483c-aec5-ebac4e0bd7a4/execute/sync; $sData={"script":"return document.readyState", "args":[]}
__WD_Post: URL=HTTP://127.0.0.1:4444/session/f8e62cdd-0833-483c-aec5-ebac4e0bd7a4/execute/sync
__WD_Post ==> Success [0] : HTTP status = 200 ResponseText={"value":"complete"}
_WD_ExecuteScript ==> Success [0]
_WD_LoadWait ==> Success [0]
__WD_Post: URL=HTTP://127.0.0.1:4444/session/f8e62cdd-0833-483c-aec5-ebac4e0bd7a4/element; $sData={"using":"xpath","value":""}
__WD_Post: URL=HTTP://127.0.0.1:4444/session/f8e62cdd-0833-483c-aec5-ebac4e0bd7a4/element
__WD_Post ==> Invalid Expression [11] : HTTP status = 400 ResponseText={"value":{"error":"invalid selector","message":"Given xpath expression \"\" is invalid: SyntaxError: Document.evaluate: The expression is not a legal expression","stacktrace":"WebDriverError@chrome://remote/content/shared/webdriver/Errors.jsm:183:5\nInvalidSelectorError@chrome://remote/content/shared/webdriver/Errors.jsm:343:5\nfind_@chrome://remote/content/marionette/element.js:328:11\nelement.find/</findElements<@chrome://remote/content/marionette/element.js:282:24\nevalFn@chrome://remote/content/marionette/sync.js:134:7\nPollPromise/<@chrome://remote/content/marionette/sync.js:154:5\nPollPromise@chrome://remote/content/marionette/sync.js:125:10\nelement.find/<@chrome://remote/content/marionette/element.js:280:24\nelement.find@chrome://remote/content/marionette/element.js:279:10\nfindElement@chrome://remote/content/marionette/actors/MarionetteCommandsChild.jsm:243:20\nreceiveMessage@chrome://remote/content/marionette/actors/MarionetteCommandsChild.jsm:99:31\n"}}
_WD_FindElement ==> Invalid Expression [11] : Parameters:   Strategy=xpath   Selector=
__WD_Post: URL=HTTP://127.0.0.1:4444/session/f8e62cdd-0833-483c-aec5-ebac4e0bd7a4/element; $sData={"using":"css selector","value":""}
__WD_Post: URL=HTTP://127.0.0.1:4444/session/f8e62cdd-0833-483c-aec5-ebac4e0bd7a4/element
__WD_Post ==> Invalid Expression [11] : HTTP status = 400 ResponseText={"value":{"error":"invalid selector","message":"Given css selector expression \"\" is invalid: InvalidSelectorError: Document.querySelector: '' is not a valid selector: \"\"","stacktrace":"WebDriverError@chrome://remote/content/shared/webdriver/Errors.jsm:183:5\nInvalidSelectorError@chrome://remote/content/shared/webdriver/Errors.jsm:343:5\nfind_@chrome://remote/content/marionette/element.js:328:11\nelement.find/</findElements<@chrome://remote/content/marionette/element.js:282:24\nevalFn@chrome://remote/content/marionette/sync.js:134:7\nPollPromise/<@chrome://remote/content/marionette/sync.js:154:5\nPollPromise@chrome://remote/content/marionette/sync.js:125:10\nelement.find/<@chrome://remote/content/marionette/element.js:280:24\nelement.find@chrome://remote/content/marionette/element.js:279:10\nfindElement@chrome://remote/content/marionette/actors/MarionetteCommandsChild.jsm:243:20\nreceiveMessage@chrome://remote/content/marionette/actors/MarionetteCommandsChild.jsm:99:31\n"}}
_WD_FindElement ==> Invalid Expression [11] : Parameters:   Strategy=css selector   Selector=
+ wd_demo.au3: Finished: UserTesting
```

@mlipok IMO the revises `Info` logging doesn't include enough details. I recall that you suggested elsewhere to show parameters (like in `_WD_FindElement` above). Other option would be to restore logging of ResponseText.

**Comment by mlipok** [comment]
> I recall that you suggested elsewhere to show parameters (like in `_WD_FindElement` above).

I need to review all other functions in order to do that.
It will take some more time but this is on my list starting from now.

> Other option would be to restore logging of ResponseText.

I would not like to do this, because then it will no longer make sense to distinguish between "Info" and "Full"

**Comment by Danp2** [comment]
> I would not like to do this, because then it will no longer make sense to distinguish between "Info" and "Full"

Actually, that isn't accurate. `Full` show both of these lines --

```
__WD_Post: URL=HTTP://127.0.0.1:4444/session/f8e62cdd-0833-483c-aec5-ebac4e0bd7a4/url; $sData={"url":"https://www.google.com"}
__WD_Post: URL=HTTP://127.0.0.1:4444/session/f8e62cdd-0833-483c-aec5-ebac4e0bd7a4/url
```

`Info` only shows the second one. This makes me question why both are shown for `Full`. 🤔

**Comment by mlipok** [comment]
after:
[UserTesting() improved](https://github.com/Danp2/au3WebDriver/pull/290/commits/cc7f880ecd2beb0a581e8b5ced555bec5c99a9fd)
[$sParameters + __WD_MessageCreator()](https://github.com/Danp2/au3WebDriver/pull/290/commits/eb54eb6067b58e056c7ead27c6b3f6d6c428089f)
[$sParameters + _WD_WaitElement show only once](https://github.com/Danp2/au3WebDriver/pull/290/commits/4a5e9e13b82a99cd70f917b0e9a63e1732cb8134)


I get:

**NEW Info:**
```
+ wd_demo.au3: Running: UserTesting
__WD_Post: URL=HTTP://127.0.0.1:4444/session/c844f59a-6ba2-4c87-b5fc-e4d8b6e3f4ff/url
__WD_Post ==> Success [0] : HTTP status = 200
_WD_Navigate ==> Success [0]
__WD_Post: URL=HTTP://127.0.0.1:4444/session/c844f59a-6ba2-4c87-b5fc-e4d8b6e3f4ff/execute/sync
__WD_Post ==> Success [0] : HTTP status = 200
_WD_ExecuteScript ==> Success [0]
_WD_LoadWait ==> Success [0]
- Test 1:
__WD_Post: URL=HTTP://127.0.0.1:4444/session/c844f59a-6ba2-4c87-b5fc-e4d8b6e3f4ff/element
__WD_Post ==> Invalid Expression [11] : HTTP status = 400
_WD_FindElement ==> Invalid Expression [11] : Parameters:   Strategy=xpath   Selector=   StartNodeID=Default   Multiple=Default   ShadowRoot=Default
- Test 2:
__WD_Post: URL=HTTP://127.0.0.1:4444/session/c844f59a-6ba2-4c87-b5fc-e4d8b6e3f4ff/element
__WD_Post ==> No match [8] : HTTP status = 404
_WD_FindElement ==> No match [8] : Parameters:   Strategy=css selector   Selector=#fake   StartNodeID=Default   Multiple=Default   ShadowRoot=Default
_WD_WaitElement ==> Success [0] : Parameters:   Delay=1000   Timeout=3000   Options=8
+ wd_demo.au3: Finished: UserTesting
```

**NEW Full**

```
+ wd_demo.au3: Running: UserTesting
__WD_Post: URL=HTTP://127.0.0.1:4444/session/38012f42-f382-461b-a843-eb0df265e3bc/url; $sData={"url":"https://www.google.com"}
__WD_Post ==> Success [0] : HTTP status = 200 ResponseText={"value":null}
_WD_Navigate ==> Success [0]
__WD_Post: URL=HTTP://127.0.0.1:4444/session/38012f42-f382-461b-a843-eb0df265e3bc/execute/sync; $sData={"script":"return document.readyState", "args":[]}
__WD_Post ==> Success [0] : HTTP status = 200 ResponseText={"value":"complete"}
_WD_ExecuteScript ==> Success [0]
_WD_LoadWait ==> Success [0]
- Test 1:
__WD_Post: URL=HTTP://127.0.0.1:4444/session/38012f42-f382-461b-a843-eb0df265e3bc/element; $sData={"using":"xpath","value":""}
__WD_Post ==> Invalid Expression [11] : HTTP status = 400 ResponseText={"value":{"error":"invalid selector","message":"Given xpath expression \"\" is invalid: SyntaxError: Document.evaluate: The expression is not a legal expression","stacktrace":"WebDriverError@chrome://remote/content/shared/webdriver/Errors.jsm:183:5\nInvalidSelectorError@chrome://remote/content/shared/webdriver/Errors.jsm:343:5\nfind_@chrome://remote/content/marionette/element.js:328:11\nelement.find/</findElements<@chrome://remote/content/marionette/element.js:282:24\nevalFn@chrome://remote/content/marionette/sync.js:134:7\nPollPromise/<@chrome://remote/content/marionette/sync.js:154:5\nPollPromise@chrome://remote/content/marionette/sync.js:125:10\nelement.find/<@chrome://remote/content/marionette/element.js:280:24\nelement.find@chrome://remote/content/marionette/element.js:279:10\nfindElement@chrome://remote/content/marionette/actors/MarionetteCommandsChild.jsm:243:20\nreceiveMessage@chrome://remote/content/marionette/actors/MarionetteCommandsChild.jsm:99:31\n"}}
_WD_FindElement ==> Invalid Expression [11] : Parameters:   Strategy=xpath   Selector=   StartNodeID=Default   Multiple=Default   ShadowRoot=Default
- Test 2:
__WD_Post: URL=HTTP://127.0.0.1:4444/session/38012f42-f382-461b-a843-eb0df265e3bc/element; $sData={"using":"css selector","value":"#fake"}
__WD_Post ==> No match [8] : HTTP status = 404 ResponseText={"value":{"error":"no such element","message":"Unable to locate element: #fake","stacktrace":"WebDriverError@chrome://remote/content/shared/webdriver/Errors.jsm:183:5\nNoSuchElementError@chrome://remote/content/shared/webdriver/Errors.jsm:395:5\nelement.find/</<@chrome://remote/content/marionette/element.js:300:16\n"}}
_WD_FindElement ==> No match [8] : Parameters:   Strategy=css selector   Selector=#fake   StartNodeID=Default   Multiple=Default   ShadowRoot=Default
_WD_WaitElement ==> Success [0] : Parameters:   Delay=1000   Timeout=3000   Options=8
+ wd_demo.au3: Finished: UserTesting
```

**Comment by mlipok** [comment]
as for me in 
**NEW Info**
this part:
```
__WD_Post: URL=HTTP://127.0.0.1:4444/session/c844f59a-6ba2-4c87-b5fc-e4d8b6e3f4ff/execute/sync
```
it is redundant.

**Comment by mlipok** [comment]
If URL will be only in **NEW Full**

Then:

**NEW Error**
```
+ wd_demo.au3: Running: UserTesting
- Test 1:
__WD_Post ==> Invalid Expression [11] : HTTP status = 400
_WD_FindElement ==> Invalid Expression [11] : Parameters:   Strategy=xpath   Selector=   StartNodeID=Default   Multiple=Default   ShadowRoot=Default
- Test 2:
__WD_Post ==> No match [8] : HTTP status = 404
_WD_FindElement ==> No match [8] : Parameters:   Strategy=css selector   Selector=#fake   StartNodeID=Default   Multiple=Default   ShadowRoot=Default
+ wd_demo.au3: Finished: UserTesting
```

**NEW Info:**
```
+ wd_demo.au3: Running: UserTesting
__WD_Post ==> Success [0] : HTTP status = 200
_WD_Navigate ==> Success [0]
__WD_Post ==> Success [0] : HTTP status = 200
_WD_ExecuteScript ==> Success [0]
_WD_LoadWait ==> Success [0]
- Test 1:
__WD_Post ==> Invalid Expression [11] : HTTP status = 400
_WD_FindElement ==> Invalid Expression [11] : Parameters:   Strategy=xpath   Selector=   StartNodeID=Default   Multiple=Default   ShadowRoot=Default
- Test 2:
__WD_Post ==> No match [8] : HTTP status = 404
_WD_FindElement ==> No match [8] : Parameters:   Strategy=css selector   Selector=#fake   StartNodeID=Default   Multiple=Default   ShadowRoot=Default
_WD_WaitElement ==> Success [0] : Parameters:   Delay=1000   Timeout=3000   Options=8
+ wd_demo.au3: Finished: UserTesting
```

**NEW Full**

```
+ wd_demo.au3: Running: UserTesting
__WD_Post: URL=HTTP://127.0.0.1:4444/session/38012f42-f382-461b-a843-eb0df265e3bc/url; $sData={"url":"https://www.google.com"}
__WD_Post ==> Success [0] : HTTP status = 200 ResponseText={"value":null}
_WD_Navigate ==> Success [0]
__WD_Post: URL=HTTP://127.0.0.1:4444/session/38012f42-f382-461b-a843-eb0df265e3bc/execute/sync; $sData={"script":"return document.readyState", "args":[]}
__WD_Post ==> Success [0] : HTTP status = 200 ResponseText={"value":"complete"}
_WD_ExecuteScript ==> Success [0]
_WD_LoadWait ==> Success [0]
- Test 1:
__WD_Post: URL=HTTP://127.0.0.1:4444/session/38012f42-f382-461b-a843-eb0df265e3bc/element; $sData={"using":"xpath","value":""}
__WD_Post ==> Invalid Expression [11] : HTTP status = 400 ResponseText={"value":{"error":"invalid selector","message":"Given xpath expression \"\" is invalid: SyntaxError: Document.evaluate: The expression is not a legal expression","stacktrace":"WebDriverError@chrome://remote/content/shared/webdriver/Errors.jsm:183:5\nInvalidSelectorError@chrome://remote/content/shared/webdriver/Errors.jsm:343:5\nfind_@chrome://remote/content/marionette/element.js:328:11\nelement.find/</findElements<@chrome://remote/content/marionette/element.js:282:24\nevalFn@chrome://remote/content/marionette/sync.js:134:7\nPollPromise/<@chrome://remote/content/marionette/sync.js:154:5\nPollPromise@chrome://remote/content/marionette/sync.js:125:10\nelement.find/<@chrome://remote/content/marionette/element.js:280:24\nelement.find@chrome://remote/content/marionette/element.js:279:10\nfindElement@chrome://remote/content/marionette/actors/MarionetteCommandsChild.jsm:243:20\nreceiveMessage@chrome://remote/content/marionette/actors/MarionetteCommandsChild.jsm:99:31\n"}}
_WD_FindElement ==> Invalid Expression [11] : Parameters:   Strategy=xpath   Selector=   StartNodeID=Default   Multiple=Default   ShadowRoot=Default
- Test 2:
__WD_Post: URL=HTTP://127.0.0.1:4444/session/38012f42-f382-461b-a843-eb0df265e3bc/element; $sData={"using":"css selector","value":"#fake"}
__WD_Post ==> No match [8] : HTTP status = 404 ResponseText={"value":{"error":"no such element","message":"Unable to locate element: #fake","stacktrace":"WebDriverError@chrome://remote/content/shared/webdriver/Errors.jsm:183:5\nNoSuchElementError@chrome://remote/content/shared/webdriver/Errors.jsm:395:5\nelement.find/</<@chrome://remote/content/marionette/element.js:300:16\n"}}
_WD_FindElement ==> No match [8] : Parameters:   Strategy=css selector   Selector=#fake   StartNodeID=Default   Multiple=Default   ShadowRoot=Default
_WD_WaitElement ==> Success [0] : Parameters:   Delay=1000   Timeout=3000   Options=8
+ wd_demo.au3: Finished: UserTesting
```

**Comment by mlipok** [comment]
hm...  also wondering if  **NEW Error**

should 
```
+ wd_demo.au3: Running: UserTesting
- Test 1:
__WD_Post ==> Invalid Expression [11] : HTTP status = 400
_WD_FindElement ==> Invalid Expression [11] : Parameters:   Strategy=xpath   Selector=   StartNodeID=Default   Multiple=Default   ShadowRoot=Default
- Test 2:
__WD_Post ==> No match [8] : HTTP status = 404
_WD_FindElement ==> No match [8] : Parameters:   Strategy=css selector   Selector=#fake   StartNodeID=Default   Multiple=Default   ShadowRoot=Default
+ wd_demo.au3: Finished: UserTesting
```

or should not
```
+ wd_demo.au3: Running: UserTesting
- Test 1:
__WD_Post ==> Invalid Expression [11] : HTTP status = 400
_WD_FindElement ==> Invalid Expression [11]
- Test 2:
__WD_Post ==> No match [8] : HTTP status = 404
_WD_FindElement ==> No match [8]
+ wd_demo.au3: Finished: UserTesting
```

show parameters ?

**Comment by mlipok** [comment]
If  **NEW Error** should not show parameters in logs 
then it will be like:

**NEW Error**
```
+ wd_demo.au3: Running: UserTesting
- Test 1:
__WD_Post ==> Invalid Expression [11] : HTTP status = 400
_WD_FindElement ==> Invalid Expression [11]
- Test 2:
__WD_Post ==> No match [8] : HTTP status = 404
_WD_FindElement ==> No match [8]
+ wd_demo.au3: Finished: UserTesting
```

**NEW Info:**
```
+ wd_demo.au3: Running: UserTesting
__WD_Post ==> Success [0] : HTTP status = 200
_WD_Navigate ==> Success [0]
__WD_Post ==> Success [0] : HTTP status = 200
_WD_ExecuteScript ==> Success [0]
_WD_LoadWait ==> Success [0]
- Test 1:
__WD_Post ==> Invalid Expression [11] : HTTP status = 400
_WD_FindElement ==> Invalid Expression [11] : Parameters:   Strategy=xpath   Selector=   StartNodeID=Default   Multiple=Default   ShadowRoot=Default
- Test 2:
__WD_Post ==> No match [8] : HTTP status = 404
_WD_FindElement ==> No match [8] : Parameters:   Strategy=css selector   Selector=#fake   StartNodeID=Default   Multiple=Default   ShadowRoot=Default
_WD_WaitElement ==> Success [0] : Parameters:   Delay=1000   Timeout=3000   Options=8
+ wd_demo.au3: Finished: UserTesting
```

**NEW Full**

```
+ wd_demo.au3: Running: UserTesting
__WD_Post: URL=HTTP://127.0.0.1:4444/session/38012f42-f382-461b-a843-eb0df265e3bc/url; $sData={"url":"https://www.google.com"}
__WD_Post ==> Success [0] : HTTP status = 200 ResponseText={"value":null}
_WD_Navigate ==> Success [0]
__WD_Post: URL=HTTP://127.0.0.1:4444/session/38012f42-f382-461b-a843-eb0df265e3bc/execute/sync; $sData={"script":"return document.readyState", "args":[]}
__WD_Post ==> Success [0] : HTTP status = 200 ResponseText={"value":"complete"}
_WD_ExecuteScript ==> Success [0]
_WD_LoadWait ==> Success [0]
- Test 1:
__WD_Post: URL=HTTP://127.0.0.1:4444/session/38012f42-f382-461b-a843-eb0df265e3bc/element; $sData={"using":"xpath","value":""}
__WD_Post ==> Invalid Expression [11] : HTTP status = 400 ResponseText={"value":{"error":"invalid selector","message":"Given xpath expression \"\" is invalid: SyntaxError: Document.evaluate: The expression is not a legal expression","stacktrace":"WebDriverError@chrome://remote/content/shared/webdriver/Errors.jsm:183:5\nInvalidSelectorError@chrome://remote/content/shared/webdriver/Errors.jsm:343:5\nfind_@chrome://remote/content/marionette/element.js:328:11\nelement.find/</findElements<@chrome://remote/content/marionette/element.js:282:24\nevalFn@chrome://remote/content/marionette/sync.js:134:7\nPollPromise/<@chrome://remote/content/marionette/sync.js:154:5\nPollPromise@chrome://remote/content/marionette/sync.js:125:10\nelement.find/<@chrome://remote/content/marionette/element.js:280:24\nelement.find@chrome://remote/content/marionette/element.js:279:10\nfindElement@chrome://remote/content/marionette/actors/MarionetteCommandsChild.jsm:243:20\nreceiveMessage@chrome://remote/content/marionette/actors/MarionetteCommandsChild.jsm:99:31\n"}}
_WD_FindElement ==> Invalid Expression [11] : Parameters:   Strategy=xpath   Selector=   StartNodeID=Default   Multiple=Default   ShadowRoot=Default
- Test 2:
__WD_Post: URL=HTTP://127.0.0.1:4444/session/38012f42-f382-461b-a843-eb0df265e3bc/element; $sData={"using":"css selector","value":"#fake"}
__WD_Post ==> No match [8] : HTTP status = 404 ResponseText={"value":{"error":"no such element","message":"Unable to locate element: #fake","stacktrace":"WebDriverError@chrome://remote/content/shared/webdriver/Errors.jsm:183:5\nNoSuchElementError@chrome://remote/content/shared/webdriver/Errors.jsm:395:5\nelement.find/</<@chrome://remote/content/marionette/element.js:300:16\n"}}
_WD_FindElement ==> No match [8] : Parameters:   Strategy=css selector   Selector=#fake   StartNodeID=Default   Multiple=Default   ShadowRoot=Default
_WD_WaitElement ==> Success [0] : Parameters:   Delay=1000   Timeout=3000   Options=8
+ wd_demo.au3: Finished: UserTesting
```

**Comment by mlipok** [comment]
Summary about what I propose here is:

$_WD_DEBUG_None - do not show any errors in console
$_WD_DEBUG_Error - show only error number and their descriptions + HTTP Status for each _WD_Post/Get/Delete calls
$_WD_DEBUG_Info - show additionall information - parameters which was used in function call
$_WD_DEBUG_Full - show all information including URL and ResponseText

**Comment by Danp2** [comment]
> $_WD_DEBUG_Error - show (only when errors occurs) error number and their descriptions + HTTP Status for _WD_Post/Get/Delete calls

For best chance at identifying the problem, I would propose that this should behave like $_WD_DEBUG_Full, but only log when an error has occurred.

**Comment by mlipok** [comment]
hm.. maybe **$_WD_DEBUG_Diag** ?

**$_WD_DEBUG_None** - do not show any log in console
**$_WD_DEBUG_Error** - show **(only when errors occurs)** error number and their descriptions + HTTP Status for _WD_Post/Get/Delete calls
**$_WD_DEBUG_Info** - show **(always)** error number, description, HTTP status and **additionally**: parameters which was used in function call
**$_WD_DEBUG_Diag** - show **(only when errors occurs)** all information including URL and ResponseText
**$_WD_DEBUG_Full** - show **(always)** all information including URL and ResponseText

**Comment by mlipok** [comment]
Or
hm.. maybe **$_WD_DEBUG_Raw** ?

**$_WD_DEBUG_None** - do not show any log in console
**$_WD_DEBUG_Info** - show **(always)** error number, description, HTTP status and **additionally**: parameters which was used in function call
**$_WD_DEBUG_Raw** - show **(only when errors occurs)** error number and their descriptions + HTTP Status for _WD_Post/Get/Delete calls
**$_WD_DEBUG_Error** - show **(only when errors occurs)** all information including URL and ResponseText
**$_WD_DEBUG_Full** - show **(always)** all information including URL and ResponseText

**Comment by Danp2** [comment]
> as for me log generated by **NEW Info** in this part:
> 
> ```
> __WD_Post: URL=HTTP://127.0.0.1:4444/session/c844f59a-6ba2-4c87-b5fc-e4d8b6e3f4ff/execute/sync
> ```
> 
> it is redundant.
> 
> And should be only part of **NEW Full** log.

Agreed. I may be confused, but I thought it was previously decided that _WD_Post/Get/Delete calls would only be displayed with $_WD_DEBUG_Full.

**Comment by Danp2** [comment]
Why do we need another logging level? Can't we just do the below?

>  $_WD_DEBUG_Error - show (only when errors occurs) all information including URL and ResponseText

**Comment by mlipok** [comment]
I think that there is a need to have RAW error numbers with short descriptions without any additionall information.

**Comment by mlipok** [comment]
> Agreed. I may be confused, but I thought it was previously decided that _WD_Post/Get/Delete calls would only be displayed with $_WD_DEBUG_Full.

[Redundant URL when $_WD_DEBUG_Info](https://github.com/Danp2/au3WebDriver/pull/290/commits/5b8c2bb15ce47da1fd9324eaf7e4b689bc1db15a)

**Comment by Danp2** [comment]
I'm not convinced. I propose that we table that for the moment and make our best efforts to complete this PR. You can always open an new issue for additional discussions on the idea.

**Comment by mlipok** [comment]
I will back to this statement:
[The debug flag may be helpful, but in all honesty it’s so verbose that whenever I’ve tried to look at it, it’s been like trying to find a needle in a haystack and I’m not familiar with the output so I don’t even know what the needle looks like…](https://www.autoitscript.com/forum/topic/205553-webdriver-udf-help-support-iii/?do=findComment&comment=1500003)

The intention was to give simple error log at first.
Then some more informational.
Eventually full log.

> Summary about what I propose here is:
> 
> **$_WD_DEBUG_None** - do not show any log in console 
> **$_WD_DEBUG_Error** - show **(only when errors occurs)** error number and their descriptions + HTTP Status for _WD_Post/Get/Delete calls
> **$_WD_DEBUG_Info** - show **(always)** error number, description, HTTP status and **additionally:** parameters which was used in function call
> **$_WD_DEBUG_Full** - show **(always)** all information including URL and ResponseText


> > $_WD_DEBUG_Error - show (only when errors occurs) error number and their descriptions + HTTP Status for _WD_Post/Get/Delete calls
> 
> For best chance at identifying the problem, I would propose that this should behave like $_WD_DEBUG_Full, but only log when an error has occurred.

`For best change at identifying` users should switch from **Error** to **Info** (currently default) or event to **Full**
Please notice that we was trimming `current form of FULL/FULL log` to smaller amount of information, I can say to more human readable (not advance user).

**Comment by mlipok** [comment]
so if we put response text to `$_WD_DEBUG_Error` then we back ... not moving forward.

**Comment by mlipok** [comment]
btw. in some cases my current proposal about 
**$_WD_DEBUG_Error** - show **(only when errors occurs)** error number and their descriptions + HTTP Status for _WD_Post/Get/Delete calls

is more GDPR compliant, because in this form there is no way to put LoginName, Passwords, Personal ID to console.

**Comment by mlipok** [comment]
Using recently commited:

With this testing code:
```autoit
Func UserTesting() ; here you can replace the code to test your stuff before you ask on the forum
	_WD_Navigate($sSession, 'https://github.com/Danp2/au3WebDriver/pull/316')
	_WD_LoadWait($sSession)
	Local $sElement

	; success
	$sElement = _WD_WaitElement($sSession, $_WD_LOCATOR_ByCSSSelector, '#repository-container-header')
	ConsoleWrite("! ---> @error=" & @error & "  @extended=" & @extended & @CRLF)

	; timeout
	$sElement = _WD_WaitElement($sSession, $_WD_LOCATOR_ByCSSSelector, '#repository-container-header2')
	ConsoleWrite("! ---> @error=" & @error & "  @extended=" & @extended & @CRLF)

	; not valid selector
	$sElement = _WD_WaitElement($sSession, $_WD_LOCATOR_ByCSSSelector, '#')
	ConsoleWrite("! ---> @error=" & @error & "  @extended=" & @extended & @CRLF)

	; invalid session
	$sElement = _WD_WaitElement('', $_WD_LOCATOR_ByCSSSelector, '#')
	ConsoleWrite("! ---> @error=" & @error & "  @extended=" & @extended & @CRLF)

	Return $sElement

EndFunc   ;==>UserTesting
```

I get such results:
```
+ wd_demo.au3: Running: UserTesting
__WD_Post ==> Success [0] : HTTP status = 200
_WD_Navigate ==> Success [0]
__WD_Post ==> Success [0] : HTTP status = 200
_WD_ExecuteScript ==> Success [0]
_WD_LoadWait ==> Success [0]
_WD_WaitElement ==> Success [0] : Parameters:   Delay=Default   Timeout=Default   Options=Default
! ---> @error=0  @extended=0
_WD_WaitElement ==> Timeout [7] : Parameters:   Delay=Default   Timeout=Default   Options=Default
! ---> @error=7  @extended=0
_WD_WaitElement ==> Invalid Expression [11] : Parameters:   Delay=Default   Timeout=Default   Options=Default
! ---> @error=11  @extended=0
_WD_WaitElement ==> Unknown Command [16] : Parameters:   Delay=Default   Timeout=Default   Options=Default
! ---> @error=16  @extended=0
+ wd_demo.au3: Finished: UserTesting
````
Normaly ([previously without #290](https://github.com/Danp2/au3WebDriver/pull/290)) all this following errors
```
_WD_WaitElement ==> Timeout [7] : Parameters:   Delay=Default   Timeout=Default   Options=Default
_WD_WaitElement ==> Invalid Expression [11] : Parameters:   Delay=Default   Timeout=Default   Options=Default
_WD_WaitElement ==> Unknown Command [16] : Parameters:   Delay=Default   Timeout=Default   Options=Default
```
will shown a ton of logs from `_WD_FindElement` which was not needed at all.

**Comment by Danp2** [comment]
I would like to discuss this proposed code from _WD_ElementAction --

https://github.com/Danp2/au3WebDriver/blob/74986cdb24646e8d8395b011bb51de46f6eb85a5/wd_core.au3#L729-L731

I would suggest an alternative where an alternative value is shown (ie: `<masked>`). This would demonstrate that a value was passed to $sOption without displaying the actual contents.

**Comment by mlipok** [comment]
Agree.
Please provide commit

**Comment by Danp2** [comment]
Done. See https://github.com/Danp2/au3WebDriver/pull/290/commits/f78e30d6da71ea87ff2727aaef31a63800ed330b

**Comment by mlipok** [comment]
Very well
ps.
Typing on smarphone, for this reason I request this commit from you.

**Comment by Danp2** [comment]
Oh... I had assumed you were just lazy. 🤣
